### PR TITLE
feature: #98 | Ability to recover from specific exceptions

### PR DESCRIPTION
**Pull request by a-anand-91119**
@ShivamNagpal What you are about to read is entirely from LLM. I wasted some time reading this to make sure it correct wrt to the change. You also spend some time reading this 😉 

Thoughts on this? Really need this exception specific recovered and handlers. 

------------
  # Add Exception-Type Filtering to `recover()` and `onFailure()` Methods

  ## Summary

  This PR enhances the `Try` class with exception-type filtering capabilities for both recovery and error handling operations. Users can now handle multiple exception types with a single handler, making error handling more concise and maintainable.

  ## Changes

  ### 1. Varargs `recover()` Method
  Added a new `recover` method that accepts multiple exception types:

  ```java
  @SafeVarargs
  public final <E extends Exception> Try<T> recover(
      Function<? super Exception, ? extends Try<T>> failureMapper,
      Class<? extends E>... exceptionTypes
  )
```
  Example Usage:
  ```java
  Try.to(() -> readFile())
      .recover(
          ex -> Try.success("default content"),
          IOException.class,
          FileNotFoundException.class,
          SocketTimeoutException.class
      );
```
###  2. Type-Specific `onFailure()` Methods

  Added two new onFailure overloads for conditional side effects:

  Single Exception Type:
  ```java
  public <E extends Exception> Try<T> onFailure(
      Class<? extends E> exceptionType,
      Consumer<? super Exception> errorConsumer
  )
```
  Multiple Exception Types:
  ```java
  @SafeVarargs
  public final <E extends Exception> Try<T> onFailure(
      Consumer<? super Exception> errorConsumer,
      Class<? extends E>... exceptionTypes
  )
```
  Example Usage:
  ```java
  Try.to(() -> fetchData())
      .onFailure(IOException.class, ex -> logger.error("I/O error: {}", ex.getMessage()))
      .onFailure(TimeoutException.class, ex -> logger.warn("Timeout: {}", ex.getMessage()))
      .onFailure(ex -> logger.error("Unexpected error", ex));
```
  
###  Parameter Ordering

  The methods use different parameter orders to avoid method overloading ambiguity:

  | Method                     | Signature                     | Reasoning                            |
  |----------------------------|-------------------------------|--------------------------------------|
  | Single exception recover   | recover(Class, Function)      | Exception type first reads naturally |
  | Varargs recover            | recover(Function, Class...)   | Function first prevents ambiguity    |
  | Single exception onFailure | onFailure(Class, Consumer)    | Exception type first for consistency |
  | Varargs onFailure          | onFailure(Consumer, Class...) | Consumer first prevents ambiguity    |

  This design prevents Java compiler ambiguity while optimizing ergonomics for the most common case (single exception type).

 ### Why Varargs for recover?

  Before:
  ```java
  Try.to(() -> operation())
      .recover(IOException.class, ex -> Try.success("default"))
      .recover(TimeoutException.class, ex -> Try.success("default"))
      .recover(SocketException.class, ex -> Try.success("default"));
```
  After:
  ```java
  Try.to(() -> operation())
      .recover(
          ex -> Try.success("default"),
          IOException.class,
          TimeoutException.class,
          SocketException.class
      );
```
  This eliminates repetitive chaining when multiple exception types need the same recovery logic.

 ###  Why Type-Specific onFailure?

  The onFailure methods are for side effects (logging, metrics, cleanup), not transformation. Type filtering enables:

  1. Differentiated logging levels based on exception type
  2. Type-specific metrics tracking
  3. Conditional cleanup operations
  4. Chainable error observability without modifying the Try

### 1.  Predicate-Based onHandle()

  Rejected approach:
```java
  public Try<T> onHandle(
      Predicate<Try<T>> condition,
      Consumer<? super T> successConsumer,
      Consumer<? super Exception> failureConsumer
  )
```
  Why rejected:
  - compose() needs predicates for transformation logic, but onHandle() is purely for side effects
  - Exception type filtering covers 95% of conditional side-effect use cases
  - Would lead to awkward empty consumer parameters
  - Adds API complexity without sufficient value

  Users needing custom conditions can use the existing single-consumer onHandle():
```java
  .onHandle(aTry -> {
      if (customCondition(aTry)) {
          handleIt(aTry);
      }
  })
```
 ### 2.  Removing Single Exception recover() Method

  Rejected approach: Only keep the varargs version

  Why rejected:
  - Single exception recovery is the most common case (~80-90%)
  - Having Class parameter first reads more naturally: recover(IOException.class, handler)
  - Varargs forces function-first ordering to avoid ambiguity
  - Both methods serve different ergonomic needs

 ### 3.  Reverse Parameter Order in Single Exception recover()

  Rejected approach: Make single exception match varargs order: recover(Function, Class)

  Why rejected:
  - Would cause method overloading ambiguity with varargs version
  - Java cannot distinguish recover(Function, Class) from recover(Function, Class...)
  - Code would not compile

 ## Breaking Changes

  None. All changes are additive - existing API remains unchanged.

 ## Questions for Reviewers

  1. API Design: Do the different parameter orders between single and varargs versions feel intuitive, or would you prefer a different approach?
  2. Method Naming: Should the varargs versions have different names (e.g., recoverAny(), onFailureAny()) to make the distinction clearer?
  3. Use Cases: Are there other exception handling patterns we should consider supporting?
-----------------

Thoughts?

## Summary by CodeRabbit

* **New Features**
  * Improved conditional composition to route success/failure flows based on observable runtime state.
  * Added type-specific recover and onFailure overloads (including varargs support and null-safe filtering) for finer-grained error handling.

* **Tests**
  * Expanded unit tests covering type-filtered recover/onFailure (single and varargs), null/edge cases, chaining, and real-world logging scenarios.

* **Documentation**
  * Enhanced inline docs and examples for the new overloads.

### CODE REVIEW

**Comment by a-anand-91119** [comment]
@coderabbitai review

File: src/main/java/org/zeplinko/commons/lang/ext/core/Try.java [hunk 1]
```diff
@@ -189,7 +189,7 @@ public <U> Try<U> compose(
     ) {
         Objects.requireNonNull(successMapper);
         Objects.requireNonNull(failureMapper);
-        return this.isFailure() ? failureMapper.apply(this.getError()) : successMapper.apply(this.getData());
+        return isSuccess() ? successMapper.apply(getData()) : failureMapper.apply(getError());
```
Comment (ShivamNagpal) [inline src/main/java/org/zeplinko/commons/lang/ext/core/Try.java:192]: nit: It is logically same, Earlier isFailure is being used everywhere, it will ensure we follow same thing here. Can you revert this line to original condition/

**Review by ShivamNagpal — APPROVED** [review]
